### ISSUE TITLE
Error 503 in istio connection between envoy and the app 

### ISSUE THREAD

**Issue by quinont** [istio/istio#31422]
**Bug description**
We have a 503 error in a call from one app to another with istio. we suppose that the error is between the envoy and the app. 

**Can we increase the connection timeout or idle timeout between envoy and the app?**

We have a GKE cluster (1.17.15-gke.800) with istio in version 1.7.6 (this isn’t a addon component of gke). 
In a simplified scheme of the whole cluster (which has almost 450 apps), we have 2 apps that communicate from "A" to "B".

App A -> app B 

When we make a query to a specific path of app A we are having a 503 that comes from the envoy of app A. But if we occupy another path, we do not see any error and the request works correctly.
We are only seeing this error in some artifacts and it isn’t general.
The apps are made in java 11.0.4 with spring boot framework.

When we see the logs, we see that there is a request from A to B that returns a 200. and that app A returns a 200 (this is seen in the log of app A). But the envoy from app A keeps returning a 503.

The curl is:

```
# time curl --location 'http://myapp.myns/one/path/to/win' --header 'aheader: myappheader' --header 'Content-Type: application/json' --data '{
   "data01": "thingspostdata",
   "data02": "more really important data"
}' -v

*   Trying 10.88.4.171...
* TCP_NODELAY set
* Expire in 200 ms for 4 (transfer 0x56013303df90)
* Connected to myapp.myns (10.88.4.171) port 80 (#0)
> POST /one/path/to/win HTTP/1.1
> Host: myapp.myns
> User-Agent: curl/7.64.0
> Accept: */*
> Aheader: myappheader
> Content-Type: application/json
> Content-Length: 216
>

* upload completely sent off: 216 out of 216 bytes
< HTTP/1.1 503 Service Unavailable
< content-length: 95
< content-type: text/plain
< date: Tue, 09 Mar 2021 15:20:20 GMT
< server: istio-envoy
< x-envoy-decorator-operation: myapp.myns.svc.cluster.local:80/*
<
* Connection #0 to host myapp.myns left intact

upstream connect error or disconnect/reset before headers. reset reason: connection termination
real    0m1.057s
user    0m0.006s
sys     0m0.009s
```

In the envoy log we see the following:

```
[2021-03-08T18:02:44.053Z] "POST /path/appb/to/win HTTP/1.1" 200 - "-" "-" 203 344 942 939 "-" "Java/11.0.4" "c55f30f8-db46-993a-b40f-aa0f498def48" "appb.nsappb.svc.cluster.local" "10.84.55.61:8080" outbound|80||appb.nsappb.svc.cluster.local 10.84.38.124:51182 10.88.7.44:80 10.84.38.124:55838 - -
[2021-03-08T18:02:44.031Z] "POST /one/path/to/win HTTP/1.1" 503 UC "-" "-" 216 95 966 - "-" "curl/7.64.0" "bf507dd1-d08f-4d41-abb0-aaa5e2fcba86" "myapp.myns" "127.0.0.1:8080" inbound|80|myapp|myapp.myns.svc.cluster.local 127.0.0.1:42708 10.84.38.124:8080 10.84.25.21:51208 - default
```

Note: we can implement a workaround by adding an nginx in the middle that works as a proxypass for the moment. But this is not the solution.

[ ] Docs
[ ] Installation
[ X ] Networking
[ ] Performance and Scalability
[ ] Extensions and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure
[ ] Upgrade

**Expected behavior**
We expect that 503 does not happen.

**Steps to reproduce the bug**
The bug appears only when you called a specific endpoint.

**Version** (include the output of `istioctl version --remote` and `kubectl version --short` and `helm version --short` if you used Helm)

```
client version: 1.7.6 
control plane version: 1.7.6
```

```
kubectl version --short
Server Version: v1.17.15-gke.800
```

**How was Istio installed?**
Istio was installed with istioctl

**Environment where the bug was observed (cloud vendor, OS, etc)**

The environment is gke.

Additionally, please consider running `istioctl bug-report` and attach the generated cluster-state tarball to this issue.
Refer [cluster state archive](http://istio.io/help/bugs/#generating-a-cluster-state-archive) for more details.

---

Please, if something cannot be understood or you need more information, ask me.
Sorry for my awful English...

**Comment by quinont** [comment]
The virtual service and destination route are:

```
apiVersion: networking.istio.io/v1alpha3
kind: VirtualService
metadata:
  name: myapp-vs
  namespace: myns
spec:
  hosts:
  - myapp
  http:
  - timeout: 240s
    route:
    - destination:
        host: myapp
---
kind: DestinationRule
metadata:
  name: myapp-dr
  namespace: myns
spec:
  host: myapp
  trafficPolicy:
    tls:
      mode: ISTIO_MUTUAL
  subsets:
  - name: v1
    labels:
      version: v1
    trafficPolicy:
      tls:
        mode: ISTIO_MUTUAL
```

**Comment by howardjohn** [comment]
https://karlstoney.com/2019/05/31/istio-503s-ucs-and-tcp-fun-times/ may help

**Comment by quinont** [comment]
Hello @howardjohn , 
thanks for your response, but I think that I don't have the same issue that you show in your post. 
Anyway, with your post I could to find my problem. 

I used this for the debug:
`kubectl -n myns  myapp-7d76c9dc7c-tx5n2 -c istio-proxy -- curl -XPOST -s -o /dev/null http://localhost:15000/logging?level=debug`

And I saw a lot of information, but there was a line with this particular error:
`protocol error: http/1.1 protocol error: unsupported transfer encoding`

So, I googled this I found this:
https://discuss.istio.io/t/unsupported-transfer-encoding-error-in-istio-proxy/6021

Finally, the problem was that my code forward the headers from the application that it depends (app C), and when envoy receive the response from app A, it have two same headers:
Transfer-Encoding: chunked
transfer-encoding: chunked

And this causes the envoy to send the request three times, it discarding all and returning a 503. 

So thanks again for your help, it was very useful to me. Again, sorry for my awful English. 

My bug is in my app.
